From e96435d265c5cb44195ad7c2cd1c3eec05466a28 Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Tue, 15 Jan 2019 18:44:23 +0200
Subject: [PATCH 0909/1051] net: mvneta: stop transmit idle on kernel startup

Port may be configured by Uboot to transmit IDLE, so a remote side
feels the link as UP whilst the local side never made ifconfig-up.

Stop TX on local side in probe same way as in mvneta_start/stop.

Change-Id: Icf5561641cd8877821511419c1392da2fe83eacc
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/2519
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvneta.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvneta.c b/drivers/net/ethernet/marvell/mvneta.c
index 75d6773e8f1e..86f1555e736b 100644
--- a/drivers/net/ethernet/marvell/mvneta.c
+++ b/drivers/net/ethernet/marvell/mvneta.c
@@ -4768,6 +4768,21 @@ static int mvneta_probe(struct platform_device *pdev)
 
 	platform_set_drvdata(pdev, pp->dev);
 
+	/* Port may be configured by Uboot to transmit IDLE, so a remote side
+	 * feels the link as UP. Stop TX in same way as in mvneta_start/stop.
+	 */
+	if (pp->phylink) {
+		if (rtnl_is_locked()) {
+			if (!mvneta_mdio_probe(pp))
+				mvneta_mdio_remove(pp);
+		} else {
+			rtnl_lock();
+			if (!mvneta_mdio_probe(pp))
+				mvneta_mdio_remove(pp);
+			rtnl_unlock();
+		}
+	}
+
 	if (pp->musdk_port)
 		netdev_info(dev, "Port belong to User Space (MUSDK)\n");
 
-- 
2.17.1

